Source/WebCore/platform/graphics/TrackPrivateBase.cpp

```diff
@@ -53,6 +54,76 @@ bool TrackPrivateBase::operator==(const TrackPrivateBase& track) const
         && startTimeVariance() == track.startTimeVariance();
 }
 
+void TrackPrivateBase::notifyClients(Task&& task)
+{
+    Ref sharedTask = createSharedTask<void(TrackPrivateBaseClient&)>(WTFMove(task));
+    // We ensure not to hold the lock for too long by making a copy (which are cheap)
+    // as we could potentially get a re-entrant call which would cause a deadlock.
+    Vector<ClientRecord> clients;
+    {
+        Locker locker { m_lock };
+        clients = m_clients;
+    }
+    for (auto& tuple : clients) {
+        auto& [dispatcher, weakClient, mainThread] = tuple;
+        if (dispatcher) {
+            dispatcher->get()([weakClient = WTFMove(weakClient), sharedTask] {
+                if (weakClient)
+                    sharedTask->run(*weakClient);
+            });
+        }
+    }
+}
+
+void TrackPrivateBase::notifyMainThreadClient(Task&& task)
+{
+    // There will only ever be one main thread client.
+    // We call the first one found.
+    Vector<ClientRecord> clients;
+    {
+        Locker locker { m_lock };
+        clients = m_clients;
+    }
+    for (auto& tuple : clients) {
+        auto& [dispatcher, weakClient, mainThread] = tuple;
+        if (dispatcher && mainThread) {
+            dispatcher->get()([weakClient = WTFMove(weakClient), task = WTFMove(task)] {
+                if (weakClient)
+                    task(*weakClient);
+            });
+            break;
+        }
+    }
+}
+
+size_t TrackPrivateBase::addClient(TrackPrivateBaseClient::Dispatcher&& dispatcher, TrackPrivateBaseClient& client)
+{
+    Locker locker { m_lock };
+    size_t index = m_clients.size();
+    m_clients.append(std::make_tuple(SharedDispatcher::create(WTFMove(dispatcher)), WeakPtr { client }, isMainThread()));
+    return index;
+}
+
+void TrackPrivateBase::removeClient(uint32_t index)
+{
+    Locker locker { m_lock };
+    if (m_clients.size() > index)
+        return;
+    m_clients[index] = std::make_tuple<RefPtr<SharedDispatcher>, WeakPtr<TrackPrivateBaseClient>, bool>({ }, { }, false);
+}
+
+bool TrackPrivateBase::hasClients() const
+{
+    Locker locker { m_lock };
+    return m_clients.size();
+}
+
+bool TrackPrivateBase::hasOneClient() const
```

Comment:
These kind of functions: hasClients() / hasOneClient() when locks / multithreading are involved seem a little fishy? As soon as the function returns, the answer may already be wrong (since the lock is held internally and released on return).